apps/autoscheduler/AutoSchedule.cpp

```diff
@@ -2243,16 +2285,53 @@ struct State {
             // Do all the reorders and pick which vars to
             // parallelize.
             vector<VarOrRVar> vars;
-            int64_t parallel_tasks = 1;
             vector<VarOrRVar> parallel_vars;
             bool any_parallel_vars = false, any_parallel_rvars = false;
-            for (auto it = p.second->vars.rbegin(); it != p.second->vars.rend(); it++) {
-                if (!it->exists || it->extent == 1) continue;
-                if (!it->parallel) break;
-                any_parallel_rvars |= it->var.is_rvar;
-                any_parallel_vars |= !it->var.is_rvar;
-                parallel_tasks *= it->extent;
-                parallel_vars.push_back(it->var);
+            if (!target.has_gpu_feature()) {
+                int64_t parallel_tasks = 1;
+                for (auto it = p.second->vars.rbegin(); it != p.second->vars.rend(); it++) {
+                    if (!it->exists || it->extent == 1) continue;
+                    if (!it->parallel) break;
+                    any_parallel_rvars |= it->var.is_rvar;
+                    any_parallel_vars |= !it->var.is_rvar;
+                    parallel_tasks *= it->extent;
+                    parallel_vars.push_back(it->var);
+                }
+            } else {
+                uint64_t n_parallel_loops = 0;
+                uint64_t current_parallel_loop = 0;
+
+                // Count the number of parallel loops.
+                for (auto it = p.second->vars.rbegin(); it != p.second->vars.rend(); it++) {
+                    if (!it->exists || it->extent == 1) continue;
+                    if (!it->parallel) break;
+
+                    n_parallel_loops++;
+                }
+
+                // Tag as GPUBlock and GPUThread only if we have at least two loops.
+                // We want the loop structure to be GPU blocks -> serial loops -> GPU threads.
+                // We tag the first min(3, n_parallel_loops/2) loops to be GPU blocks and
+                // tag the last min(3, n_parallel_loops/2) loops to be GPU threads.
```

Comment:
There are no guarantees on the size of anything but the innermost loop, so this looks like it would die a lot of the time. Do you have an idea of what fraction of random pipelines this successfully runs?

Reply:
(die due to using too many threads)